File: nodejs/socketengine/host.js
'use strict';

const EventEmitter = require('events').EventEmitter;
const inherits = require('util').inherits;
const ip = require('ip');

// ///////////////
// / CONSTANTS ///
// ///////////////

const IMAGE = 'image';
const NEWLINE = '\n';
const base64 = new RegExp(
    '^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$',
);
const FAMILY = 'IPv4';

const ADDR = ip.address();
const PORT = 8080;
const MAXSIZE = 1500000;
const TIMEOUT = 0;

// //////////////////////
// / CONNECTION CLASS ///
// //////////////////////

function _connection(socket, address, timeout, maxSize) {
  EventEmitter.call(this);
  this.socket = socket;
  this.address = address;
  this.timeout = timeout;
  this.maxSize = maxSize;
  this.lastData = new Date().getTime();
  this.msgBuffer = '';
  this.channels = {};
  this.listener = null;

  // ////////////////////
  // / SOCKET ACTIONS ///
  // ////////////////////

  this.socket.on('data', (bytes) => {
    this.msgBuffer += bytes.toString();
    if (this.msgBuffer != '' && this.msgBuffer != '\n') {
      const data = this.msgBuffer.split('\n');
      for (let i = 0; i < data.length; i++) {
        try {
          if (data[i] == '') {
            continue;
          }
          const msg = JSON.parse(data[i]);
          this.channels[msg['type']] = msg['data'];
          if (msg['type'] == IMAGE) {
            if (base64.test(msg['data'])) {
              this.emit(msg['type'], msg['data']);
            }
          } else {
            this.emit(msg['type'], msg['data']);
          }
          this.emit('data', msg);
        } catch (err) {}
      }
    }
  });
  this.socket.on('end', () => {
    this.emit('end');
  });

  this.socket.on('error', (err) => {
    this.emit('warning', err);
  });

  // /////////////
  // / METHODS ///
  // /////////////

  this.get = function(channel) {
    return this.channels[channel];
  };

  this.write = function(dataType, data) {
    const msg = {
      type: dataType,
      data: data,
    };
    this.socket.write(JSON.stringify(msg) + NEWLINE);
  };

  this.close = function() {
    this.socket.destroy();
  };

  // Timeout handler
  if (this.timeout > 0) {
    setInterval(() => {
      if ((new Date().getTime() - this.lastData) / 1000 > this.timeout) {
        if (this.msgBuffer == '') {
          try {
            this.reset();
          } catch (err) {}
        }
      }
    }, this.timeout);
  }
}

inherits(_connection, EventEmitter);

// ////////////////
// / HOST CLASS ///
// ////////////////

function Host(addr = ADDR, port = PORT, maxSize = MAXSIZE, timeout = TIMEOUT) {
  EventEmitter.call(this);
  this.net = require('net');

  this.addr = addr;
  this.port = port;
  this.timeout = timeout;
  this.maxSize = maxSize;

  this.clients = [];
  this.sockets = [];

  this.socketpath = {
    port: this.port,
    family: FAMILY,
    address: this.addr,
  };
  this.listener = null;
  this.opened = false;

  // ////////////
  // / SERVER ///
  // ////////////

  this.server = this.net.createServer((socket) => {
    this.listener = socket;
  });

  this.server.on('connection', (socket) => {
    for (let i = 0; i < this.sockets.length; i++) {
      if (this.sockets[i] === socket) {
        return;
      }
    }
    this.sockets.push(socket);
    const connection = new _connection(
        socket,
        socket.address(),
        this.timeout,
        this.maxSize,
    );
    this.clients.push(connection);
    connection.on('data', (msg) => {
      this.emit(msg['type'], this.get_ALL(msg['type']));
    });
    connection.on('error', (err) => {
      this.emit('warning', err);
    });
    connection.on('end', () => {
      this.emit('end');
    });
  });

  // /////////////
  // / METHODS ///
  // /////////////

  this.start = function() {
    this.server.listen(this.socketpath.port, this.socketpath.address);
    this.opened = true;
    return this;
  };

  this.get_ALL = function(channel) {
    const data = [];
    for (let i = 0; i < this.clients.length; i++) {
      const tmp = this.clients[i].get(channel);
      if (tmp != undefined) {
        data.push(tmp);
      }
    }
    return data;
  };

  this.getClients = function() {
    return this.clients;
  };

  this.write_ALL = function(channel, data) {
    for (let i = 0; i < this.clients.length; i++) {
      this.clients[i].write(channel, data);
    }
    return this;
  };

  this.close = function() {
    for (let i = 0; i < this.clients.length; i++) {
      this.clients[i].close();
    }
  };
}

inherits(Host, EventEmitter);

module.exports = exports = Host;

File: python/socketengine/host.py
import socket
from threading import Thread, Lock
from json import dumps as dictToJson
from json import loads as jsonToDict
from json.decoder import JSONDecodeError
from .common import encodeImg, generateSocket

#################
### CONSTANTS ###
#################

from .constants import ACK, NEWLINE, IMG_MSG_S, IMG_MSG_E
from .constants import ADDR, PORT, TIMEOUT, SIZE, OPEN

########################
### CONNECTION CLASS ###
########################

# pylint: disable=consider-using-enumerate
class _connection:
    def __init__(self, socketConnection, address, timeout, size):
        self.socket = socketConnection
        self.address = address
        self.canWrite = True
        self.channels = {}
        self.timeout = timeout
        self.socket.settimeout(self.timeout)
        self.size = size
        self.stopped = False
        self.lock = Lock()
        self.start()

    def start(self):
        Thread(target=self.run, args=()).start()
        return self

    def run(self):
        tmp = ''
        while True:
            if self.stopped:
                self.socket.close()
                return

            try:
                tmp += self.socket.recv(self.size).decode()
            except socket.timeout:
                continue
            except OSError:
                self.close()

            if tmp not in ('', '\n'):
                data = tmp.split('\n')
                for i in range(len(data)):
                    try:
                        msg = jsonToDict(data[i])
                        self.channels[msg['type']] = msg['data']
                        if msg['type'] == ACK:
                            self.canWrite = True
                        tmp = ''
                    except JSONDecodeError:
                        tmp = ''.join(data[i:])
                        break

    def get(self, channel):
        with self.lock:
            if channel in self.channels.keys():
                return self.channels[channel]
            return None

    def writeLock(self, channel, data):
        with self.lock:
            self.write(channel, data)

    def write(self, channel, data):
        msg = {'type': channel, 'data': data}
        self.socket.sendall(dictToJson(msg).encode() + NEWLINE)

    def writeImgLock(self, data):
        with self.lock:
            self.writeImg(data)

    def writeImg(self, data):
        if self.canWrite:
            self.canWrite = False
            msg = IMG_MSG_S + encodeImg(data) + IMG_MSG_E
            self.socket.sendall(msg + NEWLINE)

    def close(self):
        self.stopped = True


##################
### HOST CLASS ###
##################

# pylint: disable=redefined-builtin, unused-variable
class Host:
    def __init__(self, addr=ADDR, port=PORT, timeout=TIMEOUT, size=SIZE, open=OPEN):
        self.addr = addr
        self.port = port
        self.timeout = timeout
        self.size = size
        self.connections = []
        self.clients = []
        self.stopped = False
        self.opened = False
        if open:
            self.open()

    def setTimeout(self, time):
        self.timeout = time

    def open(self):
        while True:
            try:
                self.socket = generateSocket(self.timeout)
                self.socket.bind(('', self.port))
                self.socket.listen()
                break
            except OSError as error:
                raise RuntimeError('Socket address in use: {}'.format(error))
            except socket.timeout:
                continue
        self.opened = True

    def start(self):
        Thread(target=self.run, args=()).start()
        return self

    def run(self):
        tmp = ''
        while True:
            if self.stopped:
                for client in self.clients:
                    client.close()
                self.socket.close()
                return

            try:
                conn, addr = self.socket.accept()
                if addr not in self.connections:
                    self.connections.append(addr)
                    self.clients.append(_connection(conn, addr, self.timeout, self.size))
            except socket.timeout:
                continue

    def getAll(self, channel):
        data = []
        for client in self.clients:
            tmp = client.get(channel)
            if tmp is not None:
                data.append(tmp)
        return data

    def getClients(self):
        return self.clients

    def writeAllLock(self, channel, data):
        for client in self.clients:
            client.writeLock(channel, data)
        return self

    def writeAll(self, channel, data):
        for client in self.clients:
            client.write(channel, data)
        return self

    def writeImageLockAll(self, data):
        for client in self.clients:
            client.writeImg(data)
        return self

    def writeImageAll(self, data):
        for client in self.clients:
            client.writeImg(data)
        return self

    def close(self):
        self.opened = False
        for client in self.clients:
            client.close()
        self.stopped = True
